### ISSUE TITLE
Add support for systemd's systemctl reboot for passwordless reboots

### ISSUE THREAD

**Issue by matkeith** [topgrade-rs/topgrade#861]
## I want to suggest some general feature
Topgrade should allow for reboot without requiring a sudo password by using command "systemctl reboot" if the system is running systemd.

## More information
If the linux system uses systemd you do not need to reboot using sudo, which requires typing in a password. Instead use the command "systemctl reboot" which will just reboot the system without further user action

https://www.freedesktop.org/software/systemd/man/latest/sd_booted.html
"Internally, this function checks whether the directory /run/systemd/system/ exists. A simple check like this can also be implemented trivially in shell or any other language."

Seems like you can check for this folder in rust and then execute the alternate command "systemctl reboot" instead in the below function

`pub fn reboot() -> Result<()> {
    print!("Rebooting...");
    Command::new("sudo").arg("reboot").status_checked()
}`

I do not know how to test this from a side branch, as I am using bazzite, an immutable desktop. I'm willing to give it a try though if there are instructions somewhere.

**Comment by SteveLauC** [comment]
Thanks for the report!

Does the `sudo` command on your Linux box need `sudo`, from this [answer](https://askubuntu.com/a/787274/1417245), `reboot` since Ubuntu 16 does not need `sudo`, and I started using Fedora since 36, I don't remember `sudo` is ever needed. Are they due to the switch to the systemd init program?

Though the `reboot` on my macOS machine does need `sudo`.

**Comment by matkeith** [comment]
@SteveLauC yeah it looks like typing "reboot" without sudo works as well for me on bazzite as well

**Comment by SteveLauC** [comment]
I found that this root-less reboot only works if you have local access to your machine, i.e., not ssh, when you do it via ssh:

```sh
$ reboot
Call to Reboot failed: Interactive authentication required.
```

See also: [Getting errors whenever I try to reboot or shutdown through terminal](https://www.reddit.com/r/linux4noobs/comments/5vbmq8/getting_errors_whenever_i_try_to_reboot_or/)

**Comment by matkeith** [comment]
Bummer. I found some additional info from ubuntu's forum as well
https://askubuntu.com/questions/1005884/sudo-reboot-vs-reboot-what-is-the-difference

specifically these criteria:

On Ubuntu 14.10 and older, sudo is required.

The introduction of Systemd in 15.04 changed the way Ubuntu handles shutdown and reboot:

    When a single user is logged in, sudo isn't necessary. When more than one user is logged in then sudo is required.

    Applications can inhibit shutdown and reboot. You override these inhibitions with sudo.

    A single user logged in via ssh still requires sudo.

so it seems there are more nuances involved than simply using ssh. reboot would fail on a multiuser system or if an application were to prevent it.

    matkeith@mattlaptop:~$ ls -al $(which reboot)
    lrwxrwxrwx. 19 root root 16 Jul  5 15:30 /usr/sbin/reboot -> ../bin/systemctl

reboot is just pointed at systemctl reboot anyhow. I don't see a way around this intelligently. oh well =/

**Comment by SteveLauC** [comment]
So, it seems that:

* If we are running Topgrade through SSH, then `sudo` is needed
* Otherwise, check if there are more than one user is logged in, if so, sudo is also needed. 

Let me implement this.

**Comment by uwuclxdy** [comment]
I suggest (during this process, if a previous step requires auth, topgrade silently ignores and tries the next one):

1. Topgrade tries to reboot with `reboot` only.
2. Check which init system is present by reading `cat /proc/1/comm` (`require()` won't do; there can be multiple init systems installed at once, but only one running):
```
$ cat /proc/1/comm
systemd
```

3. If `systemd`, run `systemctl reboot` (this here can be expanded to support other init systems if they offer the same functionality).
4. If other, or `3.` fails, finally fallback to `sudo reboot` .

thoughts?

**Comment by GideonBear** [comment]
Ref #1299, there's a bunch of different reboot stuff going on that might relate to this

> I suggest (during this process, if a previous step requires auth, topgrade silently ignores and tries the next one):

This looks okay but kind of off-topic for Topgrade, maybe this already exists as a / should be a separate crate?
https://crates.io/search?q=reboot

**Comment by uwuclxdy** [comment]
https://crates.io/crates/system_shutdown looks promising

**Comment by GideonBear** [comment]
Ref #1299
